src/daemon/daemon.cpp

```diff
@@ -678,26 +763,48 @@ mp::Daemon::Daemon(std::unique_ptr<const DaemonConfig> the_config)
 {
     connect_rpc(daemon_rpc, *this);
     std::vector<std::string> invalid_specs;
-    bool mac_addr_missing{false};
+
     for (auto& entry : vm_instance_specs)
     {
         const auto& name = entry.first;
         auto& spec = entry.second;
 
+        // Store here the MAC addresses we need add to the daemon's set. But we'll add them only if this instance
+        // is not invalid.
+        std::unordered_set<std::string> instance_mac_addresses;
+
         if (!config->vault->has_record_for(name))
         {
             invalid_specs.push_back(name);
             continue;
         }
 
-        auto mac_addr = spec.mac_addr;
-        if (mac_addr.empty())
+        // Check MAC addresses of the default and extra interfaces. For each MAC, repetition on the instance MAC
+        // addresses and in the daemon's MAC addresses is checked. String validity was already checked in
+        // load_db().
+        if (!mpu::valid_mac_address(spec.default_interface.mac_address) ||
+            !instance_mac_addresses.insert(spec.default_interface.mac_address).second ||
+            allocated_mac_addrs.find(spec.default_interface.mac_address) != allocated_mac_addrs.end())
         {
-            mac_addr = mp::utils::generate_mac_address();
-            vm_instance_specs[name].mac_addr = mac_addr;
-            mac_addr_missing = true;
+            mpl::log(mpl::Level::warning, category,
+                     fmt::format("{} has invalid or repeated default MAC address {}", name,
+                                 spec.default_interface.mac_address));
+            invalid_specs.push_back(name);
+            continue;
+        }
+
+        for (auto& iface : spec.extra_interfaces)
+        {
+            if (!mp::utils::valid_mac_address(iface.mac_address) ||
+                !instance_mac_addresses.insert(iface.mac_address).second ||
+                allocated_mac_addrs.find(iface.mac_address) != allocated_mac_addrs.end())
+            {
+                mpl::log(mpl::Level::warning, category,
+                         fmt::format("{} has invalid or repeated extra MAC address {}", name, iface.mac_address));
+                invalid_specs.push_back(name);
+                continue;
```

Comment:
Hmm, I don't think this is what we want here. This would continue on to the next iteration of the inner loop, but then go on to create the vm corresponding to this spec, which was considered invalid. I think we'd need to `break` here and then somehow signal that to `continue` in the outer loop. 

But actually, I think this would all be easier with a separate function to put the macs in a set and use larger building blocks to check for intersection with existing macs. For instance, something like this, in pseudocode:

```c++
auto new_macs = mac_set_from(spec) // this would copy the default and extra macs to a set and return it
if(auto num_new_macs = new_macs.size(); num_new_macs <= spec.extra_interfaces.size()) // we need all extras plus the default
  // there are repeated macs in the spec, so invalid
else if (new_macs.merge(allocated_macs); new_macs.size() < allocated_macs.size() + num_new_macs)
  // we know the spec repeats macs from other instances, so invalid
else
  allocated_macs = new_macs;
  // we're good, keep going
```

Reply:
Right, didn't note the glitch with `continue`s. Your suggestion makes the code clearer, I implemented it, with some improvement. `unordered_set::merge()` splices the elements from the source, so it will become empty if all the elements could be added. For this reason, it suffices to test `allocated_macs` for emptyness.

Reply:
Cool, much nicer. And you're right about merge. Further simplification possible though, see inline in new code.